### PR TITLE
Tag editor improvement

### PR DESCRIPTION
**Pull request by tramhao**
Tag editor can fetch tag, fetch lyrics, preview lyrics, delete lyrics now.

### CODE REVIEW

**Comment by issadarkthing** [comment]
I will review this tonight

**Comment by tramhao** [comment]
The code to fetch lyric in tag editor is very ugly as I cannot do it by calling lyricPopup. The problem is, lyricPopup is embeded with go routines, and I need to update content of Dropdown and textview based on the result from lyricpopup. When doing so, my code of updaing runs before getting lyrics. I tried to use channel or waitgroup, but both method end up with totally freezing. Finally I have to copy code from lyricPopup again. Do you have some suggestions? Thanks.

**Comment by issadarkthing** [comment]
I think `lyric` package should have a `LyricOptions` struct with generalized `GetLyric() string` method to fetch a particular language based on specified `Lang` member on `LyricOptions` struct. This way, we can generalize the `lyricPopup` function to `lyricPopup(audioFile player.Audio, langExt string, serviceProvider string)` where `lyricPopupCN` is not needed anymore.

**Comment by issadarkthing** [comment]
> The problem is, lyricPopup is embeded with go routines, and I need to update content of Dropdown and textview based on the result from lyricpopup.

If `lyric` package is refactored as I suggested above, you can just use the `GetLyricOptions` method.

**Comment by issadarkthing** [comment]
> I think lyric package should have a LyricOptions struct with generalized func (l *LyricOptions) GetLyric() string method to fetch a particular language based on specified Lang member on LyricOptions struct. 

I think `GetLyric` and `GetLyricOptions` should be method for `Lyric` struct.

**Comment by tramhao** [comment]
> I think `lyric` package should have a `LyricOptions` struct with generalized `func (l *LyricOptions) GetLyric() string` method to fetch a particular language based on specified `Lang` member on `LyricOptions` struct. This way, we can generalize the `lyricPopup` function to `lyricPopup(audioFile player.Audio, langExt string, serviceProvider string)` where `lyricPopupCN` is not needed anymore.

But the problem here is, between GetLyricOptions and GetLyric, there is a manual selection step involved.

**Comment by tramhao** [comment]
> I think `lyric` package should have a `LyricOptions` struct with generalized `func (l *LyricOptions) GetLyric() string` method to fetch a particular language based on specified `Lang` member on `LyricOptions` struct. This way, we can generalize the `lyricPopup` function to `lyricPopup(audioFile player.Audio, langExt string, serviceProvider string)` where `lyricPopupCN` is not needed anymore.

Also, the code to fetch en lyric and cn lyric are different.

**Comment by tramhao** [comment]
> > The problem is, lyricPopup is embeded with go routines, and I need to update content of Dropdown and textview based on the result from lyricpopup.
> 
> If `lyric` package is refactored as I suggested above, you can just use the `GetLyricOptions` method.

Actually I'm using GetLyricOptions directly right now. It's just I have to repeat similar code for 3 buttons, and it doesn't feel good.

**Comment by tramhao** [comment]
I think we have three problems here:
1. Inside lyric package, lyric.go and lyric_chinese.go should be modified in some way, but the code inside these two files are quite different.
2. In popup.go, lyricpopup and lyricpopupCN should be merged.
3. In tageditor, the 3 buttons to fetch lyrics are very similar and need to be simplified.

**Comment by issadarkthing** [comment]
> 1. Inside lyric package, lyric.go and lyric_chinese.go should be modified in some way, but the code inside these two files are quite different.

For every language, they should be in their own file, example for english `lyricEn.go` and they need to have a function `getLyricEn() (string, error)` and `getLyricOptionsEn() (map[string]string, error)`. 

Lyric package should expose `GetLyric(lang string) string` and `GetLyricOptions(lang string) (map[string]string, error)`. 
Example of `GetLyric` implementation might be:
```go
func GetLyric(lang string) string {
    switch lang {
      case "jp": return getLyricJp()
      case "cn": return getLyricCN()
      default: return getLyricEn()
    }
}
```

This will make it easier to support more language.

**Comment by tramhao** [comment]
Yes and I've done it in the latest push. Now it feels much better.

**Comment by issadarkthing** [comment]
image

Is there any way to fix the color around border?

**Comment by issadarkthing** [comment]
> 3. In tageditor, the 3 buttons to fetch lyrics are very similar and need to be simplified.

I think language option to fetch lyric should be dropdown instead of button and below the dropdown should have the "fetch" button.

File: lyric/lyric.go [hunk 1]
```diff
@@ -1,64 +1,51 @@
...
+type SongTag struct {
+	Artist          string
+	Title           string
+	Album           string
+	URL             string
+	TitleForPopup   string
+	LangExt         string
+	ServiceProvider string
+	SongID          string // SongID and LyricID is returned by cn server. It's not guaranteed to be identical
+	LyridID         string
```
Comment (issadarkthing) [inline lyric/lyric.go:18]: there is typo here

**Comment by tramhao** [comment]
> image
> 
> Is there any way to fix the color around border? This is more visible on desktop with light background.

This is an issue with chinese letters displayed when border padding of flex is set to 0.
pic-full-210309-1349-09

You can check the left side and it's annoying. 

When border padding is set to 1:
pic-full-210309-1351-56

It seems all right.

**Comment by issadarkthing** [comment]
I mean the transparent color around the line. Not the border padding itself. I didn't notice this before since I was using dark desktop background.

**Comment by tramhao** [comment]
> I mean the transparent color around the line. Not the border padding itself. I didn't notice this before since I was using dark desktop background.

Probably this pr in tview is related

https://github.com/rivo/tview/pull/523

**Comment by issadarkthing** [comment]
this pr doesn't even get merged https://github.com/rivo/tview/pull/524

**Comment by tramhao** [comment]
how about this one? seems more active

**Comment by tramhao** [comment]
https://gitlab.com/tslocum/cview/-/commit/1af0da7606b8476944b5740bb4f0b711aaf2a1df

**Comment by issadarkthing** [comment]
We will try converting library to https://gitlab.com/tslocum/cview on a new pr. Let's just focus on merging current pr to the master branch.

**Comment by issadarkthing** [comment]
Are there anything else to be done on this pr?

**Comment by tramhao** [comment]
Let me figure out the color inside dropdown in a minute.

**Comment by tramhao** [comment]
Some colors cannot be set. So nothing to change for now. Thanks.

**Comment by tramhao** [comment]
done.

**Comment by issadarkthing** [comment]
Thank you for your contribution! Looking forward for more pr from you 🚀

**Pull request merged**